cli/validate_inputs.go

```diff
@@ -0,0 +1,188 @@
+package cli
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/gruntwork-io/terragrunt/config"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+// validateTerragruntInputs will collect all the terraform variables defined in the target module, and the terragrunt
+// inputs that are configured, and compare the two to determine if there are any unused inputs or undefined required
+// inputs.
+func validateTerragruntInputs(terragruntOptions *options.TerragruntOptions, workingConfig *config.TerragruntConfig) error {
+	required, optional, err := terraformModuleVariables(terragruntOptions)
+	if err != nil {
+		return err
+	}
+	allVars := append(required, optional...)
+
+	allInputs, err := getDefinedTerragruntInputs(terragruntOptions, workingConfig)
+	if err != nil {
+		return err
+	}
+
+	// Unused variables are those that are passed in by terragrunt, but is not defined in terraform.
+	unusedVars := []string{}
+	for _, varName := range allInputs {
+		if !util.ListContainsElement(allVars, varName) {
+			unusedVars = append(unusedVars, varName)
+		}
+	}
+
+	// Missing variables are those that are required by the terraform config, but not defined in terragrunt.
+	missingVars := []string{}
+	for _, varName := range required {
+		if !util.ListContainsElement(allInputs, varName) {
+			missingVars = append(missingVars, varName)
+		}
+	}
+
+	// Now print out all the information
+	if len(unusedVars) > 0 {
+		terragruntOptions.Logger.Warn("The following inputs passed in by terragrunt are unused:\n")
+		for _, varName := range unusedVars {
+			terragruntOptions.Logger.Warnf("\t- %s", varName)
+		}
+		terragruntOptions.Logger.Warn("")
+	} else {
+		terragruntOptions.Logger.Info("All variables passed in by terragrunt are in use.")
+	}
+
+	if len(missingVars) > 0 {
+		terragruntOptions.Logger.Warn("The following required inputs are missing:\n")
+		for _, varName := range missingVars {
+			terragruntOptions.Logger.Warnf("\t- %s", varName)
+		}
+		terragruntOptions.Logger.Warn("")
+	} else {
+		terragruntOptions.Logger.Info("All required inputs are passed in by terragrunt.")
+	}
+
+	// Return an error when there are misaligned inputs.
+	if len(unusedVars) > 0 || len(missingVars) > 0 {
+		return fmt.Errorf("Terragrunt configuration has misaligned inputs")
+	}
+	return nil
+}
+
+// getDefinedTerragruntInputs will return a list of names of all variables that are configured by terragrunt to be
+// passed into terraform. Terragrunt can pass in inputs from:
+// - var files defined on terraform.extra_arguments blocks.
```

Comment:
Hm, I think there's are several categories of variables missing here: 

- `.tfvars`, `.tfvars.json`, `.auto.tfvars`, and `*.auto.tfvars.json` files in the working directory.
- `.tfvars`, `.tfvars.json`, `.auto.tfvars`, and `*.auto.tfvars.json` files in the module code.
- Var files passed in via `-var-file` CLI args (not `extra_args`, but normal CLI args passed to the `terragrunt` command).
- Variables passed in via `-var` CLI args in `extra_args`.
- Variables passed in via `-var` CLI args (not `extra_args`, but normal CLI args passed to the `terragrunt` command).

There may be others too, which makes me worry about maintaining this functionality... Is there any way to run `terraform` to tell us which vars got passed in? E.g., Run some command with no resources, only the input vars?

Reply:
> Is there any way to run terraform to tell us which vars got passed in?

Nope there's nothing that will tell us this. Otherwise, I would have used that here :)

Reply:
> Hm, I think there's are several categories of variables missing here:

I went for the 80-20 rule here. I think what I have covers 80% of the use cases (possibly more), for 20% of the effort.

Out of those, the configured `-var` CLI args are the only one in that list IMO that probably is worth supporting given that that is a static configuration. I added that in: 6632f4c (I also added docs in that commit to clarify which variable sources are checked in the command.)

For the others, I think it is ok if we punt supporting them until someone requests it:

- I think it's an anti pattern to be using automatically loaded tfvars in a terragrunt project. It has an element of surprise, and breaks the rule where you can't use `terragrunt.hcl` as a root (that is, looking at `terragrunt.hcl` won't tell you the full picture anymore).

- I avoided the CLI args based variable manipulation (those that are passed in via normal CLI args) because they are dynamic across each run. I view this more of a linter for the terragrunt configuration, and it feels wrong to me for a linter to change its results based on dynamic CLI args.*

\* I was on the fence about including the `TF_VAR` env vars as well for the same reason, but since we have this logic deeply baked in to the variable setting logic of terragrunt, I decided to just include it.

Reply:
> > Is there any way to run terraform to tell us which vars got passed in?
> 
> Nope there's nothing that will tell us this. Otherwise, I would have used that here :)

One option to consider:

- Extract solely the variable declarations from the Terraform code into a tmp folder.
- Run `terraform plan -out=tmp.plan` in that folder.
- Run `terraform show -json tmp.plan`, passing all variables the way Terragrunt would normally do it. I believe this will give you the values Terraform found for every variable?

Reply:
Ah, perhaps the `plan` command will fail if some of the variables are not set though... Damn, so close.

Reply:
> > Hm, I think there's are several categories of variables missing here:
> 
> I went for the 80-20 rule here. I think what I have covers 80% of the use cases (possibly more), for 20% of the effort.

If we already have all the logic to handle `-var` and `-var-files` in `extra_arguments`, is it that much more to find the same passed via CLI commands and to check for the `.tfvars` files?

I'm pushing on this because I have the impression that it's not 80% of use cases... `.tfvars` files in particular seem very common in all Terraform usage, including Terragrunt, in no small part because earlier versions of Terragrunt had their configuration within `.tfvars` files; also, we didn't used to have `read_terragrunt_config`, `yamldecode`, etc, so reliance on `.tfvars` files was very common!

Reply:
Ohhh... One more thought on using Terraform to validate the args:

- As of 0.12.x, if you set a variable _in a `.tfvars` file_ (not an env var) that isn't defined in a variable block, Terraform would give an error.
- In all versions, if you didn't set a required variable, Terraform would give an error.

So I wonder if this new command should write all Terragrunt variables to a temporary `.tfvars` file, run `plan -json` or `validate -json` with that, delete the temporary file, and report the results?

Reply:
> If we already have all the logic to handle -var and -var-files in extra_arguments, is it that much more to find the same passed via CLI commands and to check for the .tfvars files?

Ok I will add this in, because you are right: this is probably trivial to add now. That said...

> I'm pushing on this because I have the impression that it's not 80% of use cases... .tfvars files in particular seem very common in all Terraform usage, including Terragrunt, in no small part because earlier versions of Terragrunt had their configuration within .tfvars files; also, we didn't used to have read_terragrunt_config, yamldecode, etc, so reliance on .tfvars files was very common!

I might be missing something, but I thought the tfvars usage that we've been recommending and showing in examples are all based on `extra_arguments`, and not dynamically passed args via the CLI for each run. I thought one of the benefits of terragrunt was that you **didn't** need to set those variables each time you ran `apply`/`plan`, so that it was all captured as code.

E.g., the old way (prior to `yamldecode`) of doing things used `extra_arguments` which this supports already: https://github.com/gruntwork-io/terragrunt-infrastructure-live-example/blob/7cdce0f96c580fc7808b191c5ed3fab463d81958/prod/terraform.tfvars#L19

> So I wonder if this new command should write all Terragrunt variables to a temporary .tfvars file, run plan -json or validate -json with that, delete the temporary file, and report the results?

I don't think this works. `validate` doesn't check variables (from the help text, bold mine):

> Validate runs checks that verify whether a configuration is syntactically valid and internally consistent, **regardless of any provided variables** or existing state.

And there is no `-json` option for `plan`, and AFAIK `out` doesn't produce an output if there is an error.

Reply:
> I might be missing something, but I thought the tfvars usage that we've been recommending and showing in examples are all based on `extra_arguments`, and not dynamically passed args via the CLI for each run. I thought one of the benefits of terragrunt was that you **didn't** need to set those variables each time you ran `apply`/`plan`, so that it was all captured as code.

Terragrunt deliberately copies things in pwd to the working dir, including `.tfvars` files. That was specifically so that `terraform.tfvars`, which is what we used to use for Terragrunt's config, would be copied too, and any variables set within it would "just work." I suspect there a non-trivial number of users relying on that copying behavior still.

> > So I wonder if this new command should write all Terragrunt variables to a temporary .tfvars file, run plan -json or validate -json with that, delete the temporary file, and report the results?
> 
> I don't think this works. `validate` doesn't check variables (from the help text, bold mine):
> 
> > Validate runs checks that verify whether a configuration is syntactically valid and internally consistent, **regardless of any provided variables** or existing state.
> 
> And there is no `-json` option for `plan`, and AFAIK `out` doesn't produce an output if there is an error.

Ah, bummer.

Reply:
I just did a check with `plan -out` and `validate`:

- For plan, when there is an unused var, it only shows a warning in the log, and that var doesn't show up in the plan file:

```
Warning: Value for undeclared variable

The root module does not declare a variable named "unused_input" but a value
was found in file "terraform.tfvars". To use this value, add a "variable"
block to the configuration.

Using a variables file to set an undeclared variable is deprecated and will
become an error in a future release. If you wish to provide certain "global"
settings to all configurations in your organization, use TF_VAR_...
environment variables to set these instead.


Warning: Value for undeclared variable

The root module does not declare a variable named "foo" but a value was found
in file "terraform.tfvars". To use this value, add a "variable" block to the
configuration.

Using a variables file to set an undeclared variable is deprecated and will
become an error in a future release. If you wish to provide certain "global"
settings to all configurations in your organization, use TF_VAR_...
environment variables to set these instead.
```

- When there is a missing required var, this becomes an error log and no plan file is generated:

```
Error: No value for required variable

  on main.tf line 1:
   1: variable "input" {}

The root module input variable "input" is not set, and has no default value.
Use a -var or -var-file command line argument to provide a value for this
variable.
```

- Validate doesn't fail with either error case

---

So in theory we can run plan in the mode you specified to handle this. There are a few downsides to this though, one of them being major:

- [major] Writing secrets to disk: https://github.com/gruntwork-io/terragrunt/pull/1267#issuecomment-667530771

- In the current implementation, the variables values you provide can be mocks even if you are using data sources, because it doesn't actually run any of the resources. With the `plan` approach, you can get extraneous error messages resulting from unresolvable data sources.

Reply:
Yea... If we were *nix only, we could probably generate purely in-memory `.tfvars` files, and pipe those to Terraform... But not sure there's portable way to do that. Sadly, prob not worth it.

Reply:
Ok added in autoloaded tfvars logic: ed37094